test/e2e/middleware-custom-matchers-basepath/test/index.test.ts

```diff
@@ -0,0 +1,45 @@
+/* eslint-env jest */
+
+import { join } from 'path'
+import webdriver from 'next-webdriver'
+import { fetchViaHTTP } from 'next-test-utils'
+import { createNext, FileRef } from 'e2e-utils'
+import { NextInstance } from 'test/lib/next-modes/base'
+
+describe('Middleware custom matchers basePath', () => {
+  let next: NextInstance
+
+  beforeAll(async () => {
+    next = await createNext({
+      files: new FileRef(join(__dirname, '../app')),
+    })
+  })
+  afterAll(() => next.destroy())
+
+  it.each(['/docs/hello', '/docs/about'])('should match', async (path) => {
+    const res = await fetchViaHTTP(next.url, path)
+    expect(res.status).toBe(200)
+    expect(res.headers.get('x-from-middleware')).toBeDefined()
+  })
+
+  it.each(['/hello', '/about', '/invalid/docs/hello'])(
+    'should not match',
+    async (path) => {
+      const res = await fetchViaHTTP(next.url, path)
+      expect(res.status).toBe(404)
+    }
+  )
+
+  it.each(['hello', 'about'])(
```

Comment:
Can we also test the `_next/data` route variant of these are working as expected? 

The matcher needs to automatically handle this.

Reply:
I think this test is for `/_next/data` actually since when client navigation happens, the route is triggered, or do you mean something different?

Reply:
Mainly meant explicitly requesting to `_next/data` to ensure it's matching as expected so we can check the headers directly.

Reply:
Seems it's possible to do for `next dev` but I don't know how to get the hash on other cases. Since it's tested on client navigation, I think it's ok